cpp/src/io/parquet/experimental/dictionary_page_filter.cu

```diff
@@ -46,11 +47,1472 @@ using parquet::detail::ColumnChunkDesc;
 using parquet::detail::decode_error;
 using parquet::detail::PageInfo;
 
+namespace {
+
+namespace cg = cooperative_groups;
+
+// Decode kernel parameters
+auto constexpr int96_size        = 12;   ///< Size of INT96 physical type
+auto constexpr decode_block_size = 128;  ///< Must be a multiple of warp_size
+/// Maximum number of literals to evaluate while decoding column dictionaries
+auto constexpr max_inline_literals = 2;
+
+// cuCollections static set parameters
+using key_type = cudf::size_type;                  ///< Using column indices (size_type) as set keys
+auto constexpr empty_key_sentinel = key_type{-1};  ///< We will never encounter a -1 index.
+auto constexpr set_cg_size        = 1;             ///< Cooperative group size for cuco::static_set
+auto constexpr bucket_size        = 1;             ///< Number of buckets per set slot
+auto constexpr occupancy_factor = 70;  ///< cuCollections suggests targeting a 70% occupancy factor
+uint32_t constexpr hasher_seed  = 0;   ///< default seed for the hashers
+
+/// Hash function for cuco::static_set_ref
+template <typename T>
+using hasher_type = cudf::hashing::detail::MurmurHash3_x86_32<T>;
+
+/// cuco::static_set_ref storage type
+using storage_type     = cuco::bucket_storage<key_type,
+                                          bucket_size,
+                                          cuco::extent<std::size_t>,
+                                          cudf::detail::cuco_allocator<char>>;
+using storage_ref_type = typename storage_type::ref_type;
+using bucket_type      = typename storage_type::bucket_type;
+
+/// hash set operation type
+enum class set_operation_type : uint8_t { INSERT, QUERY };
+
+/**
+ * @brief Hash functor for inserting values into `cuco::static_set`
+ *
+ * @tparam T Underlying data type of the cudf column
+ * @tparam op Type of `cuco::static_set` operation
+ */
+template <typename T, set_operation_type op>
+struct hash_functor {
+  cudf::device_span<T const> const decoded_data;
+  uint32_t const seed{hasher_seed};
+  static constexpr auto is_insert = op == set_operation_type::INSERT;
+  static constexpr auto is_query  = op == set_operation_type::QUERY;
+
+  /**
+   * @brief Hasher operator for inserting values into `cuco::static_set`
+   */
+  template <bool is_insert = is_insert>
+  constexpr __device__ __forceinline__
+    cuda::std::enable_if_t<is_insert, typename hasher_type<T>::result_type>
```

Comment:
This looks fishy. We need this because it's possible that the query type `T` can be the same as `key_type`?